stan/math/rev/functor/cvodes_integrator_adjoint.hpp

```diff
@@ -0,0 +1,731 @@
+#include <stan/math/rev/functor/cvodes_utils.hpp>
+#include <stan/math/rev/functor/coupled_ode_system.hpp>
+#include <stan/math/rev/functor/ode_store_sensitivities.hpp>
+#include <stan/math/prim/err.hpp>
+#include <stan/math/prim/fun/value_of.hpp>
+#include <cvodes/cvodes.h>
+#include <sunlinsol/sunlinsol_dense.h>
+#include <algorithm>
+#include <ostream>
+#include <utility>
+#include <vector>
+
+namespace stan {
+namespace math {
+
+template <typename T_Return>
+inline std::vector<Eigen::Matrix<T_Return, Eigen::Dynamic, 1>> build_varis(
+    vari* ode_vari, vari**& non_chaining_varis,
+    const std::vector<Eigen::VectorXd>& y);
+
+template <>
+inline std::vector<Eigen::Matrix<var, Eigen::Dynamic, 1>> build_varis<var>(
+    vari* ode_vari, vari**& non_chaining_varis,
+    const std::vector<Eigen::VectorXd>& y) {
+  std::vector<Eigen::Matrix<var, Eigen::Dynamic, 1>> y_return(y.size());
+
+  if (y.size() == 0) {
+    return y_return;
+  }
+
+  int N = y[0].size();
+
+  non_chaining_varis
+      = ChainableStack::instance_->memalloc_.alloc_array<vari*>(y.size() * N);
+
+  for (size_t i = 0; i < y.size(); ++i) {
+    for (size_t j = 0; j < N; j++) {
+      non_chaining_varis[i * N + j] = new vari(y[i][j], false);
+    }
+  }
+
+  for (size_t i = 0; i < y.size(); ++i) {
+    y_return[i].resize(N);
+    for (size_t j = 0; j < N; j++)
+      y_return[i][j] = var(non_chaining_varis[i * N + j]);
+  }
+
+  return y_return;
+}
+
+/*
+ * If theta and y are both doubles, just pass the values through (there's
+ * no autodiff to handle here).
+ */
+template <>
+inline std::vector<Eigen::VectorXd> build_varis<double>(
+    vari* ode_vari, vari**& non_chaining_varis,
+    const std::vector<Eigen::VectorXd>& y) {
+  return y;
+}
+
+template <int Lmm, typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args>
+class cvodes_integrator_adjoint_vari;
+
+template <int Lmm, typename F, typename T_y0, typename T_t0, typename T_ts,
+          typename... T_Args>
+class cvodes_integrator_adjoint_memory : public chainable_alloc {
+  using T_Return = return_type_t<T_y0, T_t0, T_ts, T_Args...>;
+  using T_y0_t0 = return_type_t<T_y0, T_t0>;
+
+  const size_t N_;
+  const F f_;
+  const Eigen::Matrix<T_y0_t0, Eigen::Dynamic, 1> y0_;
+  const T_t0 t0_;
+  const std::vector<T_ts> ts_;
+  // std::tuple<T_Args...> args_tuple_;
+  std::tuple<decltype(deep_copy_vars(std::declval<const T_Args&>()))...>
+      local_args_tuple_;
+  std::tuple<plain_type_t<decltype(value_of(std::declval<const T_Args&>()))>...>
+      value_of_args_tuple_;
+  std::vector<Eigen::VectorXd> y_;
+  void* cvodes_mem_;
+  Eigen::VectorXd state;
+  N_Vector nv_state_;
+  SUNMatrix A_;
+  SUNLinearSolver LS_;
+
+  template <require_eigen_col_vector_t<T_y0>* = nullptr>
+  cvodes_integrator_adjoint_memory(const F& f, const T_y0& y0, const T_t0& t0,
+                                   const std::vector<T_ts>& ts,
+                                   const T_Args&... args)
```

Comment:
You gotta do a weird trick here to deduce from the class. You can only do SFINAE template deduction on template parameters  _deduced_ in the template signature. But! You can just cheat and tell the compiler a default value like the below. 

```cpp
template <T_y0_checker = T_y0, require_eigen_col_vector_t<T_y0_checker>* = nullptr>
```

But tbh I'd just do

```suggestion
  template <template T_y, require_eigen_col_vector_t<T_y>* = nullptr,
    require_same_t<T_y, T_y0>* = nullptr>
  cvodes_integrator_adjoint_memory(const F& f, const T_y& y0, const T_t0& t0,
                                   const std::vector<T_ts>& ts,
                                   const T_Args&... args)
```